agents/agent.py

```diff
@@ -0,0 +1,129 @@
+    except socket.error:
+        return "UNKNOWN"
+
+
+def get_ip(hostname):
+    try:
+        return socket.gethostbyname(hostname)
+    except (socket.gaierror, socket.error):
+        return "UNKNOWN"
+
+
+def get_linux_uptime():
+    with open("/proc/uptime", "r") as f:
+        return float(f.readline().split()[0])
+
+
+def get_uptime(os_type):
+    if "Linux" in os_type:
+        return get_linux_uptime()
+
+    return "UNKNOWN"
+
+
+def get_timestamp():
+    return datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
+
+
+class ServerAgent(object):
+    """
+    Base class for all agents. Handles operations common for all
+    servers, such as getting server metadata and writing it to logfile.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, logfile):
+        self.logfile = logfile
+
+        self.os_type = platform.system() or "UNKNOWN"
+
+        self.hostname = get_hostname()
+        self.ip = get_ip(self.hostname)
+
+        self.uptime = get_uptime(self.os_type)
+        self.timestamp = get_timestamp()
+
+        # Subclass must set port
+        self.port = -1
+
+    def port_open(self):
+        """
+        Check if the port is open.
+
+        Returns:
+            bool: Port status.
+
+        Raises:
+            ValueError: If the port not assigned a valid number.
+        """
+        if self.port == -1:
+            raise ValueError(
+                "Port not set: server agent must assign a valid port number"
+            )
+
+        # Set a TCP/IP socket
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        try:
+            s.settimeout(2)
+            s.connect((
+                self.hostname if self.hostname != "UNKNOWN" else "localhost",
+                self.port,
+            ))
+        except socket.error:
+            return False
+        finally:
+            s.close()
+
+        return True
+
+    @abc.abstractmethod
+    def service_healthy(self):
+        """
+        Check if the specific service (SMTP, DNS, etc.) is running and
+        healthy.
+        """
+        pass
+
+    def to_dict(self):
+        return {
+            "os": self.os_type,
+            "hostname": self.hostname,
+            "ip": self.ip,
+            "uptime": self.uptime,
+            "timestamp": self.timestamp,
+            "healthy": self.service_healthy(),
+        }
+
+    def to_json(self):
+        """Dump host metadata to json file."""
+        try:
+            with open(self.logfile, "w") as f:
+                json.dump(self.to_dict(), f)
+        except (IOError, OSError) as e:
+            logger.error("Error writing to %s: %s" % (self.logfile, e))
+
+    def to_txt(self):
+        """Dump host metadata to txt file as key-value pairs."""
+        data = self.to_dict()
+
+        try:
+            with open(self.logfile, "w") as f:
+                for k, v in data.items():
+                    f.write((u"%s: %s\n" % (k, v)).encode("utf-8"))
```

Comment:
While this works in Python 2, it's better to handle encoding at file-open level using:
```
import codecs
with codecs.open(self.logfile, "w", encoding="utf-8") as f:
```

Reply:
I fully agree that setting encoding for each line is not the best idea. But since I switched to using a logger, there is no explicit file-opening anymore.